lib/decode.js

```diff
@@ -0,0 +1,143 @@
+/*!
+ * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
+ */
+import {CborldError} from './error';
+import cbor from 'cbor';
+import {getCborldContextUrls} from './context';
+import {getTermCodecMap} from './codec';
+import util from 'util';
+
+/**
+ * Decodes a CBOR-LD byte array into a JSON-LD document.
+ *
+ * @param {object} options - The options to use when decoding CBOR-LD.
+ * @param {Uint8Array} options.cborldBytes - The encoded CBOR-LD bytes to
+ *   decode.
+ * @param {Function} options.documentLoader -The document loader to use when
+ *   resolving JSON-LD Context URLs.
+ * @param {Map} [options.appContextMap] - A map of JSON-LD Context URLs and
+ *   their associated CBOR-LD values. The values must be greater than
+ *   32767 (0x7FFF)).
+ * @param {diagnosticFunction} [options.diagnose] - A function that, if
+ *   provided, is called with diagnostic information.
+ *
+ * @returns {object} - The decoded JSON-LD Document.
+ */
+export async function fromCborld({
+  cborldBytes, documentLoader, appContextMap, diagnose}) {
+  const cborMap = cbor.decode(cborldBytes);
+
+  let jsonldDocument = {};
+  if(cborMap.tag === 0x0500) {
+    jsonldDocument = cborMap.value;
+  } else if(cborMap.tag === 0x0501) {
+    const contextUrls =
+      getCborldContextUrls({cborMap: cborMap.value, appContextMap});
+    const appTermMap = cborMap.value.get(0);
+    cborMap.value.delete(0);
+    const termCodecMap = await getTermCodecMap(
+      {contextUrls, appTermMap, documentLoader, decode: true});
+
+    // generate the decoding Map from the CBOR map
+    const decodingMap = _generateDecodingMap({
+      cborMap: cborMap.value, appContextMap, appTermMap, termCodecMap});
+
+    jsonldDocument = _decodeMap(decodingMap);
+  } else {
+    throw new CborldError('ERR_NOT_CBORLD',
+      'Input is not valid CBOR-LD; missing CBOR-LD header (0x0500 or 0x501).');
+  }
+
+  if(diagnose) {
+    diagnose('Diagnostic CBOR Object:');
+    diagnose(await util.inspect(cborMap, {depth: null, colors: true}));
+    diagnose('Diagnostic JSON-LD Object:');
+    diagnose(await util.inspect(jsonldDocument, {depth: null, colors: true}));
+  }
```

Comment:
`diagnose` seems more like a log function ... perhaps the docs should reflect this. Not sure what to think about this pattern -- but not going to say it should hold up this PR.

Reply:
Raised as issue https://github.com/digitalbazaar/cborld/issues/5.